### ISSUE TITLE
Realm: bogus footprint value returned on allocation failure

### ISSUE THREAD

**Issue by fdouglis** [StanfordLegion/legion#905]
Due to operator error, I was invoking FlexFlow with too small a GPU memory allocation.  It was failing with a message that it was trying to allocate 0 bytes.  I traced this to the FlexFlow custom mapper, which was not passing &footprint into the memory allocation function or the error function, hence the default of 0.  But when I modified it to follow the same design pattern as the default_mapper, it would report that it was allocating an enormous region -- it would round to 128TB, but this value was not identical from run to run, so it looks like it was returning garbage.  

Here is a context diff to flexflow that adds the footprint to the calls:

[mapper.patch.txt](https://github.com/StanfordLegion/legion/files/4996618/mapper.patch.txt)

May well be more operator error in my attempt to model the behavior of default_mapper, or it may be a true bug.

**Comment by fdouglis** [comment]
Actually, let me update this to say that with this change, there appear to be valid allocations that also get bogus footprints -- not just the case where it failed.  Also, I wondered if "%lu" for the size_t was possibly an issue; I changed it to use a portable #defined code and type, with no effect.

**Comment by lightsighter** [comment]
Can you say which commit you are using? Also do you have a pointer to your code? If not that is ok too. I'm mainly try to see which of the instance creation calls in the mapper runtime you are using:
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_mapping.h#L1828-1863

**Comment by fdouglis** [comment]
Hi Mike, thanks.  I am using the version of Legion that is embedded in FlexFlow.  I had originally simply taken the latest FlexFlow to build in a Docker but I realized that between one build and the next things broke due to inconsistencies.  So I picked one that seemed to run ok, `24c1454627c4a5d9fcd4cb7a26f2ae4e7b5d326d`.  I then manually copy in at the moment 3 files that overwrite that build -- each change is basically adding log statements only, with the exception of the change in the patch I attached above which also adds a parameter to call in and get the "footprint" when calling from the FlexFlow mapper instead of default_mapper.  

My theory is that the path in the function it calls sets *footprint before various returns, but there is at least some path where it is not doing so, and that's leaving it initialized to whatever was in the stack when it allocated the footprint variable.  I could probably confirm that by setting it to some special value and seeing if it's still that value when it returns ...

**Comment by lightsighter** [comment]
Can you say which of those four functions I linked above are being called in the FlexFlow mapper? Also, is this a multi-node run or a single node run? It would be helpful if you could confirm that the data ends up being uninitialized or whether garbage data is being written there.

**Comment by fdouglis** [comment]
Not sure what is being called -- FlexFlow is calling DefaultMapper::default_create_custom_instances, and that in turn calls default_create_instance, which calls other stuff. etc.  I can see if I can step through it in an example that is giving garbage.  Let me get back to you shortly.

**Comment by fdouglis** [comment]
OK, it took a bit to get access to play with this again.  And I can't reproduce it, at least at the moment.  To build on a new system, I realized I needed to do the same thing for Legion as I'd done with FlexFlow -- pinning to a commit.  But when I did that, I no longer was seeing the bogus value, at least so far.  One theory is that the versions of FlexFlow and Legion I was building in docker had some inconsistency -- I see since it doesn't specify stable, it's hitting a moving target.

Let's put this on hold for now.  I'll close it, but if I see it again and can reliably reproduce, I'll reopen.

**Comment by lightsighter** [comment]
Ok, sounds good. I can also imagine there is/were bugs in the default mapper passing that pointer down to Legion. If you see symptoms of it again please reopen the issue.

**Comment by fdouglis** [comment]
OK, I've been able to reproduce it on two systems, with the same dockerfile, assuming that I invoke inception (the FlexFlow example program) with appropriate arguments for that system.  I changed the source to initialize footprint to a specific number and then I get lots of output statements as follows:

`Allocated 1234567890 bytes to proc 1d00000000000005`

So what is happening here is that when the custom mapper calls the default mapper and gets success, its footprint hasn't been updated.  I'm not sure I have seen another case recently where it fails in that call and prints out the usual message prior to abort, which used to be 0 (since the footprint wasn't passed at all).  

The specific versions I am using could be themselves inconsistent in some fashion, and I just happened to get unlucky.  I had done a checkout in docker, and it used that cached version for some time, so I stuck with that particular branch to reproduce the problem.  Maybe I caught it on a bad day.  Here is that part of the dockerfile:

> RUN cd /usr/local/FlexFlow  && \
   git submodule init && \
   git submodule update -r && \
   git checkout 24c1454627c4a5d9fcd4cb7a26f2ae4e7b5d326d  # specific branch that seems to build


> ENV FF_HOME=/usr/local/FlexFlow

> RUN cd /usr/local/FlexFlow/protobuf  && \
    ./autogen.sh && \
    ./configure && \
    make

> RUN cd /usr/local/FlexFlow/legion && \
   git checkout 16661dc3db037d66abf87801fc1c3962e8e7bc32 # this is the version we started working with .. avoid moving target

**Comment by lightsighter** [comment]
So it looks like there is one place in the default mapper where it is not passing through the `footprint` pointer to the runtime:
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/mappers/default_mapper.cc#L2235-2236
Do you think you can check to see if your code is hitting that path? I'll push a fix to the master branch of Legion, but it probably will be a little bit before it makes it into the branch and commit of Legion that FlexFlow is using.

**Comment by lightsighter** [comment]
Here is the relevant commit:
https://gitlab.com/StanfordLegion/legion/-/commit/9dbf85792e68c5ad2b1fd14033049537d87b3415#cfd528db7f8a2b6fe7a6f74d9ba667dc311d4fb2_2237_2238

**Comment by fdouglis** [comment]
(Didn't notice this until this afternoon. Hopefully now I will get alerts on updates to this issue.)  

I tried manually applying the change in that commit, but from what I can tell, it is still reporting the bogus value.  Could be operator error on my manual patch, other differences in the versions, or who knows.  I'll poke around some more.

**Comment by lightsighter** [comment]
If possible I would like to determine whether this problem is in the mapper or in the runtime. Do you think you can instrument the actual call to the runtime to see whether the value coming out of the runtime isn't correct? Also, can say whether you are running with a single process or multiple processes?

**Comment by fdouglis** [comment]
I've tried to trace through it, but it's too hard ... many instances of functions with the same name, where single-stepping just doesn't suffice, but setting breakpoints doesn't necessarily keep the same flow.

I certainly see examples where I could imagine footprint not being set.  As an example:

> 2085	      if (!target_memory.exists())
> 2086	        return false;
> 2087	      if (regions.empty())
> 2088	        return false;

It looks here like it will pop back up the stack, never having set footprint (I can see at the point where it does these tests, *footprint is the special 1234567890 value I initialized it to).  I haven't actually witnessed it returning at this point.  

I am running `alexnet -e 10 -b 16 --lr 0.1 --wd 1e-4 -p 10 -ll:gpu 1 -ll:fsize 6000 -ll:zsize 10000 -ll:cpu 4 -ll:csize 60000   -level 2,mapper=0,inst=0,event=3,amhandler=3,bgwork=3` in this case.   I assume this is one multithreaded process, but I don't know what FlexFlow is doing under the covers.

**Comment by lightsighter** [comment]
@jiazhihao Can you tell me whether this is a single node or multi-node run?

@fdouglis Is this instance allocation succeeding or failing when you get back the bad footprint size? I agree those "return false" cases in the default mapper are not good if you've asked for the footprint size, but if we're getting back a successful instance allocation then those paths aren't the source of your problem.

**Comment by fdouglis** [comment]
@lightsighter -- first, just to clarify, first you talked about multiple processes, then multiple nodes.  I'm curious if there is a distinction here.  I figured with gpu:1 and cpu:4, it is potentially multiple processes, but the system only knows about the host (container) it's on, so it's not multiple nodes, unless you're thinking of each cpu instance as a node. 

In the examples I gave there, the allocation was succeeding.  Originally, I wondered why when an allocation failed, it was reporting it asked for 0; this was because it was never even getting the footprint so it was defaulting to 0 in the error message. Then when I saw it was still failing under some scenarios, I wanted to know what was successfully allocated, so I could see if the out of memory error was legit.  I added something to get the footprint whether it was successful or not, but added a debug statement to print it in the successful path as well.  That is (at least primarily) where I see the value not being set.  I would have to play with it a bit more and see if I can tweak the failing path to see if there is a case where that happens too.

As for "the source of my problem" ... at the moment, this is a "this ought to work as expected" thing and not "this failing to work is causing me issues" -- unless the inability to get real counts of sizes is a problem.  I realize there are some other paths such as in _inst_ that may give me that data, and which I didn't know about when I added these debug statements.

**Comment by lightsighter** [comment]
I think I figured out the issue and pushed a fix with a4bb911f8 to the master branch. It will be in the control replication branch that FlexFlow uses later today. @jiazhihao will have to do update the FlexFlow submodule though to pick it up at some point.

**Comment by jiazhihao** [comment]
I have update the Legion submodule in FlexFlow to include the fix. We can close the issue if @fdouglis verifies that the bug has been fixed.

**Comment by fdouglis** [comment]
I started to build it yesterday, but then before I got back to it, we lost power in our lab.  So I'll keep you posted.  

My build probably failed anyway... I was thinking that somehow flexflow was getting the latest legion directly from github, but I guess you have your own copy, so the cherry-pick I tried throwing into Dockerfile wouldn't have worked.   I can copy it in manually of course.  

I think I had explicitly picked a commit in both FlexFlow and Legion to pin the versions down, since things were moving a bit quickly otherwise.  If I remove the checkout and take the latest, it'll probably fix this bug, but I wonder what else I'll turn up ... 

More when our lab is back up.
